File: examples/python/claims/create_professional_claim_submission_edit_failure.py
# /// script
# requires-python = ">=3.12"
# dependencies = ["stedi==1.0.0"]
# ///

"""Handle a claim that fails Stedi's edits.

    STEDI_API_KEY=<api-key> uv run --script create_professional_claim_submission_edit_failure.py

An ICD-10 code that doesn't exist trips Stedi's clearinghouse edits. The submission still
succeeds: Stedi stores the claim and a 277CA acknowledging the rejection, never sends the
claim to the payer, and returns the reasons in `errors`.

See https://www.stedi.com/docs/healthcare/claim-edits-and-repairs for more.
"""

import asyncio
import os
from pathlib import Path
from uuid import uuid4

from smithy_json import JSONCodec
from stedi import Config, Stedi
from stedi.models import CreateProfessionalClaimSubmissionInput

FIXTURE = Path(__file__).parent / "fixtures" / "test-claim.json"
INVALID_DIAGNOSIS_CODE = "FZ9888"


def load_claim_with_invalid_diagnosis() -> CreateProfessionalClaimSubmissionInput:
    """Read the sample claim and swap in a diagnosis code the edits will reject."""
    claim = JSONCodec().deserialize(FIXTURE.read_bytes(), CreateProfessionalClaimSubmissionInput)

    if not claim.payer or claim.payer.id != "STEDITEST" or not claim.billing or not claim.encounter:
        raise SystemExit("input must be complete and bill the Stedi test payer.")

    claim.idempotency_key = str(uuid4())
    claim.billing.patient_control_number = new_patient_control_number()
    claim.encounter.primary_diagnosis_code = INVALID_DIAGNOSIS_CODE
    for line in claim.service_lines or []:
        line.diagnosis_codes = [INVALID_DIAGNOSIS_CODE]
    return claim


def new_patient_control_number() -> str:
    """Your own identifier for the claim (CLM01), unique per claim; payers echo it back."""
    return f"SDKTEST-{uuid4().hex[:8].upper()}"


async def main(api_key: str) -> None:
    async with Stedi(Config(api_key=api_key)) as client:
        submission = await client.create_professional_claim_submission(
            load_claim_with_invalid_diagnosis()
        )

    errors = submission.errors or []
    if not errors:
        raise SystemExit(f"expected Stedi's edits to reject {INVALID_DIAGNOSIS_CODE}")

    print(f"claim {submission.claim_id} rejected with {len(errors)} error(s):")
    for error in errors:
        print(f"  {error.description}")


api_key = os.environ.get("STEDI_API_KEY")
if not api_key:
    raise SystemExit("STEDI_API_KEY is not set")

asyncio.run(main(api_key))

File: examples/typescript/claims/create-professional-claim-submission-edit-failure.ts
/**
 * Handle a claim that fails Stedi's edits.
 *
 *   STEDI_API_KEY=<api-key> tsx create-professional-claim-submission-edit-failure.ts
 *
 * An ICD-10 code that doesn't exist trips Stedi's clearinghouse edits. The submission still
 * succeeds: Stedi stores the claim and a 277CA acknowledging the rejection, never sends the
 * claim to the payer, and returns the reasons in `errors`.
 *
 * See https://www.stedi.com/docs/healthcare/claim-edits-and-repairs for more.
 */

import { randomUUID } from "node:crypto";
import { readFile } from "node:fs/promises";

import { Stedi, type CreateProfessionalClaimSubmissionInput } from "@stedi/sdk";

const FIXTURE = `${import.meta.dirname}/fixtures/test-claim.json`;
const INVALID_DIAGNOSIS_CODE = "FZ9888";

/** Reads the sample claim and swaps in a diagnosis code the edits will reject. */
async function loadClaimWithInvalidDiagnosis(): Promise<CreateProfessionalClaimSubmissionInput> {
  const claim = JSON.parse(
    await readFile(FIXTURE, "utf8"),
  ) as CreateProfessionalClaimSubmissionInput;

  if (claim.payer?.id !== "STEDITEST" || !claim.billing || !claim.encounter) {
    throw new Error("input must be complete and bill the Stedi test payer.");
  }

  return {
    ...claim,
    encounter: { ...claim.encounter, primaryDiagnosisCode: INVALID_DIAGNOSIS_CODE },
    serviceLines: claim.serviceLines?.map((line) => ({
      ...line,
      diagnosisCodes: [INVALID_DIAGNOSIS_CODE],
    })),
    billing: { ...claim.billing, patientControlNumber: newPatientControlNumber() },
    idempotencyKey: randomUUID(),
  };
}

/** Your own identifier for the claim (CLM01), unique per claim; payers echo it back. */
function newPatientControlNumber(): string {
  return `SDKTEST-${randomUUID().slice(0, 8).toUpperCase()}`;
}

const apiKey = process.env["STEDI_API_KEY"];
if (!apiKey) {
  throw new Error("STEDI_API_KEY is not set");
}

const client = new Stedi({ apiKey: { apiKey } });
const submission = await client.createProfessionalClaimSubmission(
  await loadClaimWithInvalidDiagnosis(),
);
const errors = submission.errors ?? [];

if (errors.length === 0) {
  throw new Error(`expected Stedi's edits to reject ${INVALID_DIAGNOSIS_CODE}`);
}

console.log(`claim ${submission.claimId} rejected with ${errors.length} error(s):`);
for (const error of errors) {
  console.log(`  ${error.description}`);
}
